cmd/containerd/command/service_windows.go

```diff
@@ -0,0 +1,467 @@
+	"time"
+	"unsafe"
+
+	"github.com/containerd/containerd/services/server"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+	"golang.org/x/sys/windows/svc/eventlog"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+var (
+	serviceNameFlag       string
+	registerServiceFlag   bool
+	unregisterServiceFlag bool
+	runServiceFlag        bool
+
+	setStdHandle = windows.NewLazySystemDLL("kernel32.dll").NewProc("SetStdHandle")
+	oldStderr    windows.Handle
+	panicFile    *os.File
+
+	service *handler
+)
+
+const (
+	// These should match the values in event_messages.mc.
+	eventInfo  = 1
+	eventWarn  = 1
+	eventError = 1
+	eventDebug = 2
+	eventPanic = 3
+	eventFatal = 4
+
+	eventExtraOffset = 10 // Add this to any event to get a string that supports extended data
+)
+
+// serviceFlags returns an array of flags for configuring containerd to run
+// as a Windows service under control of SCM.
+func serviceFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "service-name",
+			Usage: "Set the Windows service name",
+			Value: "containerd",
+		},
+		cli.BoolFlag{
+			Name:  "register-service",
+			Usage: "Register the service and exit",
+		},
+		cli.BoolFlag{
+			Name:  "unregister-service",
+			Usage: "Unregister the service and exit",
+		},
+		cli.BoolFlag{
+			Name:   "run-service",
+			Usage:  "",
+			Hidden: true,
+		},
+	}
+}
+
+// applyPlatformFlags applys platform-specific flags.
+func applyPlatformFlags(context *cli.Context) {
+
+	if s := context.GlobalString("service-name"); s != "" {
+		serviceNameFlag = s
+	}
+	for _, v := range []struct {
+		name string
+		d    *bool
+	}{
+		{
+			name: "register-service",
+			d:    &registerServiceFlag,
+		},
+		{
+			name: "unregister-service",
+			d:    &unregisterServiceFlag,
+		},
+		{
+			name: "run-service",
+			d:    &runServiceFlag,
+		},
+	} {
+		*v.d = context.GlobalBool(v.name)
+	}
+}
+
+type handler struct {
+	fromsvc chan error
+	s       *server.Server
+	done    chan struct{} // Indicates back to app main to quit
+}
+
+type etwHook struct {
+	log *eventlog.Log
+}
+
+func (h *etwHook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.PanicLevel,
+		logrus.FatalLevel,
+		logrus.ErrorLevel,
+		logrus.WarnLevel,
+		logrus.InfoLevel,
+		logrus.DebugLevel,
+	}
+}
+
+func (h *etwHook) Fire(e *logrus.Entry) error {
+	var (
+		etype uint16
+		eid   uint32
+	)
+
+	switch e.Level {
+	case logrus.PanicLevel:
+		etype = windows.EVENTLOG_ERROR_TYPE
+		eid = eventPanic
+	case logrus.FatalLevel:
+		etype = windows.EVENTLOG_ERROR_TYPE
+		eid = eventFatal
+	case logrus.ErrorLevel:
+		etype = windows.EVENTLOG_ERROR_TYPE
+		eid = eventError
+	case logrus.WarnLevel:
+		etype = windows.EVENTLOG_WARNING_TYPE
+		eid = eventWarn
+	case logrus.InfoLevel:
+		etype = windows.EVENTLOG_INFORMATION_TYPE
+		eid = eventInfo
+	case logrus.DebugLevel:
+		etype = windows.EVENTLOG_INFORMATION_TYPE
```

Comment:
INFO? This seems like a bug. We really want Debug info to be show differently ya?

Reply:
I don't think we have a lot of options here, and it's consistent with moby. Perhaps with the tracelog changes in the future, we can do better. But we're limited with the current set of levels available with simple event logging.

Reply:
Np. This is fine then